File: backend/pdf_engine.py
from __future__ import annotations

from pathlib import Path

CM_PER_INCH = 2.54
PT_PER_INCH = 72


def cm_to_pt(cm: float) -> float:
    return cm / CM_PER_INCH * PT_PER_INCH


def _fmt(n: float) -> str:
    text = f"{n:.4f}".rstrip("0").rstrip(".")
    return text or "0"


def build_pdf(jpeg_path: Path, output_path: Path, width_cm: float, height_cm: float, bleed_mm: float = 0.0) -> None:
    jpeg = jpeg_path.read_bytes()
    from PIL import Image
    with Image.open(jpeg_path) as img:
        iw, ih = img.size
    bleed_cm = bleed_mm / 10.0
    total_w_cm = width_cm + 2 * bleed_cm
    total_h_cm = height_cm + 2 * bleed_cm
    media = [0, 0, cm_to_pt(total_w_cm), cm_to_pt(total_h_cm)]
    trim = [cm_to_pt(bleed_cm), cm_to_pt(bleed_cm), cm_to_pt(bleed_cm + width_cm), cm_to_pt(bleed_cm + height_cm)]
    page_w, page_h = media[2], media[3]
    objects: dict[int, bytes] = {}
    objects[1] = b"<< /Type /Catalog /Pages 2 0 R >>"
    objects[2] = b"<< /Type /Pages /Kids [3 0 R] /Count 1 >>"
    box = lambda b: " ".join(_fmt(x) for x in b)
    objects[3] = (f"<< /Type /Page /Parent 2 0 R /MediaBox [{box(media)}] /TrimBox [{box(trim)}] /BleedBox [{box(media)}] /Resources << /XObject << /Im0 4 0 R >> >> /ProcSet [/PDF /ImageC] >> /Contents 5 0 R >>").encode()
    head = f"<< /Type /XObject /Subtype /Image /Width {iw} /Height {ih} /ColorSpace /DeviceRGB /BitsPerComponent 8 /Filter /DCTDecode /Length {len(jpeg)} >>\nstream\n".encode()
    objects[4] = head + jpeg + b"\nendstream"
    stream = f"q\n{_fmt(page_w)} 0 0 {_fmt(page_h)} 0 0 cm\n/Im0 Do\nQ\n".encode()
    objects[5] = f"<< /Length {len(stream)} >>\nstream\n".encode() + stream + b"endstream"
    objects[6] = b"<< /Title (FechaPrint AI) /Producer (FechaPrint AI Open Source Pipeline) >>"
    header = b"%PDF-1.7\n%\xe2\xe3\xcf\xd3\n"
    chunks = [header]; offsets = [0] * 7; cursor = len(header)
    for i in range(1, 7):
        offsets[i] = cursor
        chunk = f"{i} 0 obj\n".encode() + objects[i] + b"\nendobj\n"
        chunks.append(chunk); cursor += len(chunk)
    xref_offset = cursor
    xref = b"xref\n0 7\n0000000000 65535 f \n" + b"".join(f"{offsets[i]:010d} 00000 n \n".encode() for i in range(1,7))
    trailer = xref + f"trailer\n<< /Size 7 /Root 1 0 R /Info 6 0 R >>\nstartxref\n{xref_offset}\n%%EOF\n".encode()
    output_path.write_bytes(b"".join(chunks) + trailer)

File: src/pdf-engine.js
const encoder = new TextEncoder();

function ascii(str) {
  return encoder.encode(str);
}

function concat(chunks) {
  const total = chunks.reduce((sum, c) => sum + c.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    out.set(chunk, offset);
    offset += chunk.length;
  }
  return out;
}

function fmt(n) {
  return Number(n).toFixed(4).replace(/0+$/, '').replace(/\.$/, '');
}

export function buildProductionPdf({ jpegBytes, imageWidth, imageHeight, mediaBox, trimBox, bleedBox, title = 'Arquivo para impressão' }) {
  if (!(jpegBytes instanceof Uint8Array)) jpegBytes = new Uint8Array(jpegBytes);
  const [mx0, my0, mx1, my1] = mediaBox;
  const pageW = mx1 - mx0;
  const pageH = my1 - my0;

  const objects = [];
  objects[1] = ascii('<< /Type /Catalog /Pages 2 0 R >>');
  objects[2] = ascii('<< /Type /Pages /Kids [3 0 R] /Count 1 >>');
  objects[3] = ascii(`<< /Type /Page /Parent 2 0 R /MediaBox [${mediaBox.map(fmt).join(' ')}] /TrimBox [${trimBox.map(fmt).join(' ')}] /BleedBox [${bleedBox.map(fmt).join(' ')}] /Resources << /XObject << /Im0 4 0 R >> >> /ProcSet [/PDF /ImageC] >> /Contents 5 0 R >>`);
  const imageHead = ascii(`<< /Type /XObject /Subtype /Image /Width ${imageWidth} /Height ${imageHeight} /ColorSpace /DeviceRGB /BitsPerComponent 8 /Filter /DCTDecode /Length ${jpegBytes.length} >>\nstream\n`);
  objects[4] = concat([imageHead, jpegBytes, ascii('\nendstream')]);
  const stream = `q\n${fmt(pageW)} 0 0 ${fmt(pageH)} 0 0 cm\n/Im0 Do\nQ\n`;
  objects[5] = ascii(`<< /Length ${ascii(stream).length} >>\nstream\n${stream}endstream`);
  const safeTitle = String(title).replace(/[()\\]/g, '');
  objects[6] = ascii(`<< /Title (${safeTitle}) /Producer (FechaPrint AI - PDF de producao) /Creator (FechaPrint AI) >>`);

  const header = new Uint8Array([0x25,0x50,0x44,0x46,0x2d,0x31,0x2e,0x37,0x0a,0x25,0xe2,0xe3,0xcf,0xd3,0x0a]);
  const chunks = [header];
  const offsets = [0];
  let cursor = header.length;

  for (let i = 1; i <= 6; i++) {
    offsets[i] = cursor;
    const pre = ascii(`${i} 0 obj\n`);
    const post = ascii('\nendobj\n');
    chunks.push(pre, objects[i], post);
    cursor += pre.length + objects[i].length + post.length;
  }

  const xrefOffset = cursor;
  let xref = 'xref\n0 7\n0000000000 65535 f \n';
  for (let i = 1; i <= 6; i++) xref += `${String(offsets[i]).padStart(10, '0')} 00000 n \n`;
  const trailer = `${xref}trailer\n<< /Size 7 /Root 1 0 R /Info 6 0 R >>\nstartxref\n${xrefOffset}\n%%EOF\n`;
  chunks.push(ascii(trailer));
  return concat(chunks);
}
